mavlink/message.go

```diff
@@ -102,53 +103,96 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
 // a message they're interested in, and convert it to the
 // corresponding type via Message.FromPacket()
 func (dec *Decoder) Decode() (*Packet, error) {
+	startFound := false
 
-	// discard bytes until our start byte
 	for {
-		c, err := dec.br.ReadByte()
-		if err != nil {
-			return nil, err
+		// discard bytes in buffer before start byte
+		for i, b := range dec.buffer {
+			if b == startByte {
+				dec.buffer = dec.buffer[i:]
+				startFound = true
+				break
+			}
 		}
-		if c == startByte {
-			break
+
+		// if start not found, read until we see start byte
+		if !startFound {
+			for {
+				c, err := dec.br.ReadByte()
+				if err != nil {
+					return nil, err
+				}
+				if c == startByte {
+					dec.buffer = append(dec.buffer, c)
+					startFound = true
+					break
+				}
+			}
 		}
-	}
 
-	// hdr contains LENGTH, SEQ, SYSID, COMPID, MSGID
-	hdr := make([]byte, 5)
-	if _, err := io.ReadFull(dec.br, hdr); err != nil {
-		return nil, err
-	}
+		if len(dec.buffer) < 2 {
+			// read length byte
+			bytesRead := make([]byte, 1)
+			n, err := io.ReadAtLeast(dec.br, bytesRead, 1)
+			if err != nil {
+				return nil, err
+			}
 
-	p, payloadLen := newPacketFromBytes(hdr)
+			dec.buffer = append(dec.buffer, bytesRead[:n]...)
+		}
 
-	crc := x25.New()
-	crc.Write(hdr)
+		// buffer[1] is LENGTH and we've already read len(buffer) bytes
+		payloadLen := int(dec.buffer[1])
+		packetLen := hdrLen + payloadLen + numChecksumBytes
+		bytesNeeded := packetLen - len(dec.buffer)
+		if bytesNeeded > 0 {
+			bytesRead := make([]byte, bytesNeeded)
+			n, err := io.ReadAtLeast(dec.br, bytesRead, bytesNeeded)
+			if err != nil {
+				return nil, err
+			}
+			dec.buffer = append(dec.buffer, bytesRead[:n]...)
+		}
 
-	// read payload (if there is one) and checksum bytes
-	buf := make([]byte, payloadLen+numChecksumBytes)
-	if _, err := io.ReadFull(dec.br, buf); err != nil {
-		return p, err
-	}
+		// hdr contains LENGTH, SEQ, SYSID, COMPID, MSGID
+		// (hdrLen - 1) because we don't include the start byte
+		hdr := make([]byte, hdrLen-1)
+		// don't include start byte
+		hdr = dec.buffer[1:hdrLen]
 
-	p.Payload = buf[:payloadLen]
-	crc.Write(p.Payload)
+		p, payloadLen := newPacketFromBytes(hdr)
 
-	crcx, err := dec.Dialects.findCrcX(p.MsgID)
-	if err != nil {
-		return p, err
-	}
-	crc.WriteByte(crcx)
+		crc := x25.New()
+		crc.Write(hdr)
 
-	p.Checksum = bytesToU16(buf[payloadLen:])
+		payloadStart := hdrLen
+		p.Payload = dec.buffer[payloadStart : payloadStart+payloadLen]
+		crc.Write(p.Payload)
 
-	// does the transmitted checksum match our computed checksum?
-	if p.Checksum != crc.Sum16() {
-		return p, ErrCrcFail
+		crcx, err := dec.Dialects.findCrcX(p.MsgID)
+		if err != nil {
+			dec.buffer = dec.buffer[1:]
+			startFound = false
```

Comment:
The idea I had is that the caller could check the error and decide whether or not to call `Decode()` again. And `Decode()` would just continue from where it was in the buffer.

Basically if the caller sees a single `ErrUnknownMsgID` and then gets a valid packet on the next call to `Decode()` that would indicate there was possibly a problem in the stream. However, if the caller gets `ErrUnknownMsgID` on many consecutive calls, that could be a sign that the wrong dialect is being used.